infrastructure/gcp/keep-test/google-functions/keep-faucet/issue-grant.js

```diff
@@ -39,72 +38,117 @@ const web3 = new Web3(keepContractOwnerProvider, null, web3Options)
+      const start = web3.utils.toBN(Math.floor(Date.now() / 1000))
+      const cliff = web3.utils.toBN(0)
+      const revocable = false
+      const tokens = web3.utils.toBN(300000)
+      console.log(`Fetching existing balance for account [${granteeAccount}]...`)
+      const grantBalanceString = await tokenGrant.methods
+        .balanceOf(granteeAccount)
+        .call()
+      const grantBalance = web3.utils.toBN(grantBalanceString)
+      console.log(`Existing balance for account [${granteeAccount}] is [${grantBalance}].`)
+      const grantAmount = tokens.mul(tokenDecimalMultiplier)
 
-      await keepToken.methods
-        .approveAndCall(tokenGrant.address, grantAmount, grantData)
-        .send({ from: keepContractOwnerAddress })
+      if (grantBalance.gte(grantAmount)) {
+        console.warn(
+          `[${granteeAccount}] requested grant while at limit. Balance: [${grantBalance}].`,
+        )
+        return response.status(400).send(`
+          Token grant failed: your account has the maximum testnet KEEP allowed.\n
+          You can manage your token grants at: https://dashboard.test.keep.network\n
+          If you have questions, you can find us on Discord: https://discord.gg/jqxBU4m\n`
+        )
+      } else {
+        console.log(
+          `Submitting grant for [${grantAmount}] to [${granteeAccount}]...`,
+        )
+        const grantData = Buffer.concat([
+          Buffer.from(granteeAccount.substr(2), 'hex', 20),
+          unlockingDuration.toBuffer('be', 32),
+          start.toBuffer('be', 32),
+          cliff.toBuffer('be', 32),
+          Buffer.from(revocable ? '01' : '00', 'hex'),
+          Buffer.from(permissiveStakingPolicyAddress.substr(2), 'hex', 20)
+        ])
 
-      console.log(
-        `Created grant for ${web3.utils.toBN(
-          grantAmount,
-        )} to: ${granteeAccount}`,
-      )
-      response.send(`
-        Created token grant with ${web3.utils.toBN(
-          grantAmount,
-        )} KEEP for account: ${granteeAccount}
-        You can manage your token grants at: https://dashboard.test.keep.network
-        You can find us on Discord at: https://discord.gg/jqxBU4m\n`)
+        console.log("Test submission...")
+        // Try calling; if this throws, we'll have a proper error message thanks
+        // to handleRevert above.
+        await keepToken.methods
+          .approveAndCall(tokenGrantAddress, grantAmount, grantData)
+          .call({ from: keepContractOwnerAddress })
+
+        console.log("Submitting transaction...")
+        // If the call didn't revert, try submitting the transaction proper.
+        keepToken.methods
+          .approveAndCall(tokenGrantAddress, grantAmount, grantData)
+          .send({ from: keepContractOwnerAddress })
+          .on('transactionHash', (hash) => {
+            console.log(
+              `Submitted grant for [${grantAmount}] to [${granteeAccount}].`,
+              transaction,
```

Comment:
I'm not sure where `transaction` comes from here?  Should this be `hash`?

Reply:
Went ahead and tested this, the function barfs on `transaction` not being defined, but `hash` does seem to work. 

Example error.

```
Serving function...
Function: issueGrant
URL: http://localhost:8080/
Fetching existing balance for account [0xcCb2acbC6db2432d577b9EdBD42c1B125ef9925e]...
Existing balance for account [0xcCb2acbC6db2432d577b9EdBD42c1B125ef9925e] is [0].
Submitting grant for [300000000000000000000000] to [0xcCb2acbC6db2432d577b9EdBD42c1B125ef9925e]...
Test submission...
Submitting transaction...
ReferenceError: transaction is not defined
    at Promise.keepToken.methods.approveAndCall.send.on (/Users/sthompson22/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet/issue-grant.js:123:15)
    at Promise.emit (/Users/sthompson22/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet/node_modules/eventemitter3/index.js:181:35)
    at sendTxCallback (/Users/sthompson22/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet/node_modules/web3/node_modules/web3-core-method/src/index.js:625:36)
    at /Users/sthompson22/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet/node_modules/web3/node_modules/web3-core-requestmanager/src/index.js:147:9
    at /Users/sthompson22/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet/node_modules/web3-provider-engine/index.js:159:9
    at /Users/sthompson22/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet/node_modules/async/internal/once.js:12:16
    at replenish (/Users/sthompson22/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet/node_modules/async/internal/eachOfLimit.js:61:25)
    at /Users/sthompson22/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet/node_modules/async/internal/eachOfLimit.js:71:9
    at eachLimit (/Users/sthompson22/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet/node_modules/async/eachLimit.js:43:36)
    at /Users/sthompson22/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet/node_modules/async/internal/doLimit.js:9:16
    at end (/Users/sthompson22/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet/node_modules/web3-provider-engine/index.js:134:5)
    at /Users/sthompson22/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet/node_modules/async/internal/once.js:12:16
    at next (/Users/sthompson22/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet/node_modules/async/waterfall.js:21:29)
    at /Users/sthompson22/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet/node_modules/async/internal/onlyOnce.js:12:16
    at /Users/sthompson22/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet/node_modules/async/internal/once.js:12:16
    at next (/Users/sthompson22/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet/node_modules/async/waterfall.js:21:29)
```

Interesting that even after ^ is thrown the account does get a grant anyway.

Reply:
Howwww did I do that, I tested this 🤨

Reply:
Fix't.